throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,178 @@
+	if !ok {
+		return false
+	}
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID].Tier
+	blockTx := txl.contractToBlockTrx[contractAddr.String()]
+	if int64(tier.MaxTx) > blockTx.txn {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (txl *contractTxLimiter) updateState(contractAddr loom.Address, curBlockHeight int64) {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		txl.contractToBlockTrx[contractAddr.String()] = &blockTxn{1, curBlockHeight}
+		return
+	}
+	// reset blockTxn if block is out of range
+	tierID := txl.contractToTierMap[contractAddr.String()]
+	tier := txl.tierMap[tierID].Tier
+
+	if blockTx.blockHeight < curBlockHeight-int64(tier.BlockRange) {
+		blockTx.txn = 0
+		blockTx.blockHeight = curBlockHeight
+		return
+	}
+	blockTx.txn++
+}
+
+// NewContractTxLimiterMiddleware add another tx limiter that limits how many CallTx(s) can be sent to an EVM contract within a pre-configured block range
+func NewContractTxLimiterMiddleware(cfg *ContractTxLimiterConfig,
+	createUserDeployerWhitelistCtx func(state loomchain.State) (contractpb.Context, error),
+) loomchain.TxMiddlewareFunc {
+	TxLimiter = &contractTxLimiter{
+		contractToBlockTrx: make(map[string]*blockTxn, 0),
+		tierMap:            make(map[udw.TierID]tierTrack, 0),
+	}
+	return loomchain.TxMiddlewareFunc(func(
+		state loomchain.State,
+		txBytes []byte,
+		next loomchain.TxHandlerFunc,
+		isCheckTx bool,
+	) (res loomchain.TxHandlerResult, err error) {
+		if !isCheckTx {
+			return next(state, txBytes, isCheckTx)
+		}
+		var nonceTx auth.NonceTx
+		if err := proto.Unmarshal(txBytes, &nonceTx); err != nil {
+			return res, errors.Wrap(err, "throttle: unwrap nonce Tx")
+		}
+		var tx loomchain.Transaction
+		if err := proto.Unmarshal(nonceTx.Inner, &tx); err != nil {
+			return res, errors.New("throttle: unmarshal tx")
+		}
+		if tx.Id != callId {
+			return next(state, txBytes, isCheckTx)
+		}
+		var msg vm.MessageTx
+		if err := proto.Unmarshal(tx.Data, &msg); err != nil {
+			return res, errors.Wrapf(err, "unmarshal message tx %v", tx.Data)
+		}
+		var msgTx vm.CallTx
+		if err := proto.Unmarshal(msg.Data, &msgTx); err != nil {
+			return res, errors.Wrapf(err, "unmarshal call tx %v", msg.Data)
+		}
+		if msgTx.VmType == vm.VMType_EVM {
+			if TxLimiter.contractToTierMap == nil {
+				ctx, err := createUserDeployerWhitelistCtx(state)
+				if err != nil {
+					return res, errors.Wrap(err, "throttle: context creation")
+				}
+				contractToTierMap, err := udw.GetContractTierMapping(ctx)
+				if err != nil {
+					return res, errors.Wrap(err, "throttle: contractToTierMap creation")
+				}
+				TxLimiter.contractToTierMap = contractToTierMap
+				// TxLimiter.lastUpdated = time.Now().Unix()
+			}
+			contractAddr := loom.UnmarshalAddressPB(msg.To)
+			//check if contract in list
+			tierID, ok := TxLimiter.contractToTierMap[contractAddr.String()]
+			if !ok {
```

Comment:
Return from here, instead of giving an error. 

There could be contracts deployed by other means.